### DISCUSSION TITLE
BinaryPayloadBuilder.to_string type hint

### DISCUSSION THREAD

**Discussion by jamesbraza** [pymodbus-dev/pymodbus#1518] [Ideas]
`pymodbus==3.2.2`'s `BinaryPayloadBuilder.to_string` actually returns `bytes`.

Two questions:
1. Should we type hint this with `bytes`?
2. Should it be named `encode` or `to_bytes`, since it doesn't return a string?
    - Note: `BinaryPayloadBuilder.__str__` already exists, so I would think `to_string` is redundant

**Comment by janiversen** [comment]
type hinting with bytes is a good idea.

renaming to to_bytes is a good idea (remember to to add it to API_changes.rst)

__str__ is used to e.g. print the object, but should not replace a to_* method.

**Comment by janiversen** [comment]
if your preference is encode() it's ok with me.

**Comment by alexrudd2** [comment]
I agree the type hint should be bytes, but do not have a preference on `to_string` versus `encode`.

Just for information, if you set the following in `setup.cfg`
```
[mypy]
check_untyped_defs = True
```

then you can use mypy `reveal_type`.
<img width="675" alt="Screenshot 2023-04-25 at 10 00 27 AM" src="https://user-images.githubusercontent.com/52292902/234318742-353e8c1e-aa1d-469f-90d5-f24961abda17.png">
<img width="491" alt="Screenshot 2023-04-25 at 10 06 24 AM" src="https://user-images.githubusercontent.com/52292902/234320592-115cf2ee-8161-4539-9ffe-c48345e4f1b7.png">

**Comment by jamesbraza** [comment]
Okay [the PR](https://github.com/pymodbus-dev/pymodbus/pull/1526) is now merged!  One comment, is there's also `__bytes__`.

I more leave this as food for thought... it might be slick to use `__bytes__` instead of `encode`:

```python
assert builder.encode() == bytes(builder)
```

**Reply by janiversen** [reply]
no it would not, the '__' methods are magic methods describing the object or adding a specific quality.

If an app wants to manipulate the object buffer it should do so with a clear method call.

But that of course is just my opinion.
